### ISSUE TITLE
NAACL '18 | BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#14]
https://arxiv.org/pdf/1810.04805.pdf

温故知新...

**Comment by vycezhong** [comment]
现在看来，bert在nlp的地位，真的是相当于resnet在cv的地位。各种魔改层出不穷。有必要归根溯源，回顾经典！

第一个贡献是模型，证明了transformer确实牛逼，self-attention是相当于cv中卷积的东西，可以捕捉输入之间的每个词对的双向关系。

第二个贡献是输入对各个任务还通用，每个输入前加[CLS]，输入是一个sequence，一个sequence可以是一个句子或者两个句子。区分句子做了两件事：1）句子之间用[SEP]来分隔；2）segment embedding. 

第三个贡献是预训练任务，之前都是用LM，最大的缺点是只能单向训练，或者简单拼接两个方向的向量（shallow concatenation）；BERT提出了MLM和NSP。MLM就是完形填空，挖掉一些词然后预测，这里有一个问题是预测的时候是不能预测[MASK]的，但是训练的时候太多[MASK]了，所以又一个trick就是如果被选中作mask，80%标签是[MASK]，10%是一个random的词，10%不变。第二个任务是NSP，主要对QA和NLI这种需要句子之间联系的任务有用。

**Comment by vycezhong** [comment]
漏了一个更重要的贡献，就是先无监督pretrain然后再有监督finetune的模式，改变了NLP的玩法。之前很多都直接在指定任务训练集上训练，现在要基于pretrain的model。

**Comment by vycezhong** [comment]
话说有没有挖掉图片里面一些像素然后做predict的？这样图像也可以无监督训练了。不知道现在图像无监督都是怎么做的。

**Comment by yzh119** [comment]
@vycezhong ，在视觉上的应用应该很多人都想过，直接屏蔽像素大概是不work的，原因很简单，正常一张图片随意mask掉一些像素，那么这些像素直接取context的average就能恢复出来。

目前很多做法是mask掉ROI(region of interest)，去年有很多相关工作，比如[LXMERT](https://arxiv.org/pdf/1908.07490.pdf)和[VLBERT](https://arxiv.org/pdf/1908.08530.pdf)，不过都是multimodal相关的工作，直接用这种target来训练BERT for vision的应该还没有很显著的工作，今年更火的topic是self-supervised learning。

试图拿self-attention干掉CNN的工作也有：https://papers.nips.cc/paper/8302-stand-alone-self-attention-in-vision-models.pdf

**Comment by vycezhong** [comment]
纪念下第一次pretrain bert


![Screenshot-20201015085113-1410x160](https://user-images.githubusercontent.com/25879526/96407509-2091bd00-1214-11eb-8a3d-1fe35a9f3a48.png)
![Screenshot-20201015085154-888x116](https://user-images.githubusercontent.com/25879526/96407554-3dc68b80-1214-11eb-8c2f-b52e54976219.png)

**Comment by vycezhong** [comment]
https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/modeling.py#L364

self-attention计算过程
参数：[H, H] x 3 
输入: 
- hidden state:   [B, T, H] 
- attn mask: [B, h, T, T]

1. projection: [B, T, H] @ [H, H] -> [B, T, H]
2. reshape and permute: [B, T, H] -> [B, h, T, H/h] / [B, h, H/h, T]
3. compute attn scores: [B, h, T, H/h] @ [B, h, H/h, T] -> [B, h, T, T]
4. add mask: [B, h, T, T] + [B, h, T, T] -> [B, h, T, T]
5. softmax and dropout: [B, h, T, T] (dim = -1)
6. probs: [B, h, T, T] @ [B, h, T, H/h] -> [B, h, T, H/h] 
7. permute and reshape: [B, h, T, H/h] -> [B, T, H]

```py
class BertSelfAttention(nn.Module):
    def __init__(self, config):
        super(BertSelfAttention, self).__init__()
        if config.hidden_size % config.num_attention_heads != 0:
            raise ValueError(
                "The hidden size (%d) is not a multiple of the number of attention "
                "heads (%d)" % (config.hidden_size, config.num_attention_heads))
        self.num_attention_heads = config.num_attention_heads
        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
        self.all_head_size = self.num_attention_heads * self.attention_head_size

        self.query = nn.Linear(config.hidden_size, self.all_head_size)
        self.key = nn.Linear(config.hidden_size, self.all_head_size)
        self.value = nn.Linear(config.hidden_size, self.all_head_size)

        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)

    def transpose_for_scores(self, x):
        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
        x = torch.reshape(x, new_x_shape)
        return x.permute(0, 2, 1, 3)

    def transpose_key_for_scores(self, x):
        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
        x = torch.reshape(x, new_x_shape)
        return x.permute(0, 2, 3, 1)

    def forward(self, hidden_states, attention_mask):
        mixed_query_layer = self.query(hidden_states)
        mixed_key_layer = self.key(hidden_states)
        mixed_value_layer = self.value(hidden_states)

        query_layer = self.transpose_for_scores(mixed_query_layer)
        key_layer = self.transpose_key_for_scores(mixed_key_layer)
        value_layer = self.transpose_for_scores(mixed_value_layer)

        # Take the dot product between "query" and "key" to get the raw attention scores.
        attention_scores = torch.matmul(query_layer, key_layer)
        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
        # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
        attention_scores = attention_scores + attention_mask

        # Normalize the attention scores to probabilities.
        attention_probs = F.softmax(attention_scores, dim=-1)

        # This is actually dropping out entire tokens to attend to, which might
        # seem a bit unusual, but is taken from the original Transformer paper.
        attention_probs = self.dropout(attention_probs)

        context_layer = torch.matmul(attention_probs, value_layer)
        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
        new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
        context_layer = torch.reshape(context_layer, new_context_layer_shape)
        return context_layer

```

**Comment by vycezhong** [comment]
https://github.com/google-research/bert#what-is-bert

写得太好了. 终于明白了什么叫做Contextual vs context-free. 

> Pre-trained representations can also either be context-free or contextual, and contextual representations can further be unidirectional or bidirectional. Context-free models such as word2vec or GloVe generate a single "word embedding" representation for each word in the vocabulary, so bank would have the same representation in bank deposit and river bank. Contextual models instead generate a representation of each word that is based on the other words in the sentence.

> BERT was built upon recent work in pre-training contextual representations — including Semi-supervised Sequence Learning, Generative Pre-Training, ELMo, and ULMFit — but crucially these models are all unidirectional or shallowly bidirectional. This means that each word is only contextualized using the words to its left (or right). For example, in the sentence I made a bank deposit the unidirectional representation of bank is only based on I made a but not deposit. Some previous work does combine the representations from separate left-context and right-context models, but only in a "shallow" manner. BERT represents "bank" using both its left and right context — I made a ... deposit — starting from the very bottom of a deep neural network, so it is deeply bidirectional.
